### PR TITLE
Fix Gentoo Locales (SC-748)

### PR DESCRIPTION
**Pull request by holmanb**
```
Fix Gentoo locales
```

## Additional Context
Tracking down a stack trace in gentoo cloud LXD images pointed out an issue with localization on gentoo. It's possible that this code once "succeeded" with a different image build. Fix the trace and actually set a locale. 

https://wiki.gentoo.org/wiki/Localization/Guide/en

Stack Trace:
```
2022-01-24 20:04:10,553 - util.py[DEBUG]: Running module locale (<module 'cloudinit.config.cc_locale' from '/usr/lib/python3.9/site-packages/cloudinit/config/cc_locale.py'>) failed
Traceback (most recent call last):
  File "/usr/lib/python3.9/site-packages/cloudinit/stages.py", line 884, in _run_modules
    ran, _r = cc.run(run_name, mod.handle, func_args,
  File "/usr/lib/python3.9/site-packages/cloudinit/cloud.py", line 54, in run
    return self._runners.run(name, functor, args, freq, clear_on_fail)
  File "/usr/lib/python3.9/site-packages/cloudinit/helpers.py", line 186, in run
    results = functor(*args)
  File "/usr/lib/python3.9/site-packages/cloudinit/config/cc_locale.py", line 78, in handle
    cloud.distro.apply_locale(locale, locale_cfgfile)
  File "/usr/lib/python3.9/site-packages/cloudinit/distros/gentoo.py", line 43, in apply_locale
    subp.subp(['locale-gen', '-G', locale], capture=False)
  File "/usr/lib/python3.9/site-packages/cloudinit/subp.py", line 293, in subp
    raise ProcessExecutionError(stdout=out, stderr=err,
cloudinit.subp.ProcessExecutionError: Unexpected error while running command.
Command: ['locale-gen', '-G', 'en_US.UTF-8']
Exit code: 4
Reason: -
Stdout: -
Stderr: -
```

## Test Steps
1. Start gentoo cloud container (images:gentoo/cloud), note the above stack trace
2. Mount this cloud-init branch locally into container and check for error:
```
# Define BASE_DIR, NAME, and MY_CLOUD_CONFIG to appropriate values first
IMAGE_ID=images:gentoo/cloud
lxc init $IMAGE_ID $NAME -c user.user-data="$(cat $MY_CLOUD_CONFIG)"

# Gentoo currently uses 3.9 as system default 
lxc config device add $NAME host-cloud-init disk source=$BASE_DIR/cloudinit path=/usr/lib/python3.9/site-packages/cloudinit/
lxc start $NAME
lxc shell me 
```

### CODE REVIEW

**Comment by TheRealFalcon** [comment]
Is there any possibility of regressing older launches where this configuration may have been valid? Based on what I know of Gentoo, the answer is probably no, but do we know for sure?

**Comment by holmanb** [comment]
> Is there any possibility of regressing older launches where this configuration may have been valid? Based on what I know of Gentoo, the answer is probably no, but do we know for sure?

I just noticed that downstream actually has a patch on this section of code. I'll see if I can get an opinion from a distro maintainer.

**Comment by holmanb** [comment]
@prometheanfire - I see you are currently the maintainer of `app-emulation/cloud-init`. Would you mind reviewing this PR?

I noticed that [this patch](https://github.com/gentoo-mirror/gentoo/blob/stable/app-emulation/cloud-init/files/18.4-fix-filename-for-storing-locale.patch) overlaps with this section of code. My PR would introduce a change in behavior, since currently locale isn't currently set by cloud-init on Gentoo (it is set on all other distros). Would it be better to bring Gentoo in line with other distros behavior, or maintain current behavior?

File: cloudinit/distros/gentoo.py
Comment (prometheanfire) [inline cloudinit/distros/gentoo.py:41]: gentoo can use system (or not (another init)), I'm not sure how this function checks for systemd support.

File: cloudinit/distros/gentoo.py
Comment (prometheanfire) [inline cloudinit/distros/gentoo.py:66]: writing to `/etc/env.d/02locale` is not necessary if running the `eselect locale` as above.  `eselect locale` writes to `/etc/locale.conf` and `/etc/env.d/02locale` is a symlink to `/etc/locale.conf`


```
ls -l /etc/env.d/02locale
lrwxrwxrwx 14 root 2015-02-02 22:36  /etc/env.d/02locale -> ../locale.conf

cat /etc/locale.conf
# Configuration file for eselect
# This file has been automatically generated.
LANG="en_US.utf8"
```

**Review by prometheanfire — CHANGES_REQUESTED** [review]
I think some of this is good, not sure if the systemd check is needed and the file write to 02locale is can be removed

File: cloudinit/distros/gentoo.py
Comment (holmanb) [reply cloudinit/distros/gentoo.py:41]: @prometheanfire - The reason I was considering adding that is because cloud-init doesn't currently have support for systemd with Gentoo. Cloud-init assumes openrc for gentoo. I'm not sure if the ebuild packaging allows cloud-init with systemd, but if so things may fail and this log may help confused users. Alternatively we could add support for systemd to cloud-init for gentoo. I don't think it would be much work to make that happen, but I don't know if anybody would find that useful.

Comment (prometheanfire) [reply cloudinit/distros/gentoo.py:41]: `distutils-r1_python_install --init-system=sysvinit_openrc,systemd --distro gentoo` is how we are installing cloud-init, so we are at least installing service files and the like already.  Systemd on Gentoo is fully supported, even if gentoo is better known for openrc (I personally use systemd on my laptop with gentoo).

Comment (holmanb) [reply cloudinit/distros/gentoo.py:41]: @prometheanfire: I think we may be misunderstanding each other in some way.

Gentoo fully supports systemd, but cloud-init unconditionally assumes that Gentoo's init system is OpenRC.

For example, note that the [init command here](https://github.com/canonical/cloud-init/blob/6a4b6dc0de94a9e75bb44fa6c03288c7d4dded7e/cloudinit/distros/gentoo.py#L23) is rc-service, and [`apply_locale()`](https://github.com/canonical/cloud-init/blob/6a4b6dc0de94a9e75bb44fa6c03288c7d4dded7e/cloudinit/distros/gentoo.py#L35) doesn't support the [systemd locale setup with `localectl`](https://wiki.gentoo.org/wiki/Localization/Guide/en#systemd_2).

I haven't tried using cloud-init with a systemd gentoo install, but since cloud-init assumes OpenRC-specific things, I don't think it  could start services correctly.

File: cloudinit/distros/gentoo.py
Comment (holmanb) [reply cloudinit/distros/gentoo.py]: It seems I misunderstood what this part was doing, thanks. In that case I guess this might have once succeeded and set the locale. It seems that something must have changed (either in image build or `locale-gen` behavior) in some way if `locale-gen -G en_US.UTF-8` used to succeed.

Either way, I think that this adds some improvement by ensuring that `C.UTF8 UTF-8` is contained in `/etc/locale.gen` before trying to generate the local.

File: cloudinit/distros/gentoo.py
Comment (holmanb) [reply cloudinit/distros/gentoo.py:41]: @prometheanfire - Thanks for the feedback, it's much appreciated. Would you mind taking a another look at this considering my latest commit and comments above regarding systemd/cloud-init?  We have an upcoming release (22.1) and I'd like to get this in before the release if possible.

File: cloudinit/distros/gentoo.py
Comment (prometheanfire) [inline cloudinit/distros/gentoo.py:21]: being pedantic, for consistency's sake the new var should probably be suffixed with `_fn`, but I don't care too much.

**Review by prometheanfire — APPROVED** [review]

File: cloudinit/distros/gentoo.py
Comment (holmanb) [reply cloudinit/distros/gentoo.py]: Thanks!

**Review by prometheanfire — APPROVED** [review]

**Comment by holmanb** [comment]
The CI test failure before was unrelated to the change in this PR (a flakey test that relied on network connection). I rebased on main to force tests to run, no other changes. I think this is ready to merge (pending re-review by @TheRealFalcon).

**Review by TheRealFalcon — APPROVED** [review]
Thanks!

**Pull request merged**
